Simplenote/SPTextView.m

```diff
@@ -134,33 +134,67 @@ - (BOOL)checkForChecklistClick:(NSEvent *)event
 
 - (void)insertNewChecklist {
     NSRange lineRange = [self.string lineRangeForRange:self.selectedRange];
-    NSString *lineString = [self.string substringWithRange:lineRange];
+    NSUInteger cursorPosition = self.selectedRange.location;
+    NSUInteger selectionLength = self.selectedRange.length;
+    
+    // Check if cursor is at a checkbox, if so we won't adjust cursor position
+    BOOL cursorIsAtCheckbox = NO;
+    if (self.string.length >= self.selectedRange.location + 1) {
+        NSString *characterAtCursor = [self.string substringWithRange:NSMakeRange(self.selectedRange.location, 1)];
+        cursorIsAtCheckbox = [characterAtCursor isEqualToString:TextAttachmentCharacterCode];
+    }
     
+    NSString *lineString = [self.string substringWithRange:lineRange];
     BOOL didInsertCheckbox = NO;
-    if ([lineString hasPrefix:TextAttachmentCharacterCode] && [lineString length] >= 2) {
-        // Remove the checkbox
-        lineString = [lineString stringByReplacingCharactersInRange:NSMakeRange(0, 2) withString:@""];
+    NSString *resultString = @"";
+    
+    int addedCheckboxCount = 0;
+    if ([lineString hasPrefix:TextAttachmentCharacterCode] && [lineString length] >= ChecklistCursorAdjustment) {
+        // Remove the checkboxes in the selection
+        NSString *codeAndSpace = [TextAttachmentCharacterCode stringByAppendingString:@" "];
+        resultString = [lineString stringByReplacingOccurrencesOfString:codeAndSpace withString:@""];
     } else {
-        // Add a checkbox
+        // Add checkboxes to the selection
         NSString *checkboxString = [MarkdownUnchecked stringByAppendingString:@" "];
-        lineString = [checkboxString stringByAppendingString:[self.string substringWithRange:lineRange]];
+        NSArray *stringLines = [lineString componentsSeparatedByString:@"\n"];
+        for (int i=0; i < [stringLines count]; i++) {
+            NSString *line = stringLines[i];
+            // Skip any empty lines, except the first one
+            if (i != 0 && [line length] == 0) {
+                continue;
```

Comment:
I think if there's an empty line, we need to add a newline to the result string? Otherwise we lose any empty new lines when toggling a multi-line selection.